### PR TITLE
Add all configuration labels to launched slaves

### PR DESCRIPTION
**Pull request by Crazybus**
If a configuration defined multiple labels only the requiredLabel
would be added to the slave when it was created. This meant that if you
had a configuration with the labels "linux ubuntu" and started a job for
"ubuntu" it would start a new slave with only the label "ubuntu.
If you started another job for "linux" the current slave wouldn't
match so the plugin would launch a new slave. With these small changes
all of the labels will be added to the new slave.

Apart from this being unexpected I'm also fixing this so that this
plugin plays nicely with the [Single Use Slave](https://wiki.jenkins.io/display/JENKINS/Single+Use+Slave+Plugin) plugin.
All of our dynamic slaves are created with the tag "immutable" which we
are using to determine which slaves should be only used once. This means
we don't need need to keep track of which slaves are immutable and which
shouldn't be set offline after a single job.

### CODE REVIEW

**Comment by n0rad** [comment]
Can this PR be merged ? without this fix the plugin do not work as it should, start a new agent for any label combination & cannot build on multi tags jobs.

We are running on a custom build for month while the fix is available.
We updated the plugin once without the fix and screw up our CI.

**Comment by schnatterer** [comment]
This PR seems to fix [JENKINS-51277](https://issues.jenkins-ci.org/browse/JENKINS-51277).
We're also running on a custom build for the time being and desperately hope for this PR to be merged.

**Comment by rachely3n** [comment]
@craigatgoogle If you could review this one please. I will then resolve the merge conflicts.

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineCloud.java [hunk 1]
```diff
@@ -223,9 +223,9 @@ private synchronized Integer availableNodeCapacity() throws IOException {
         }
     }
 
-    private synchronized ComputeEngineInstance getAnAgent(InstanceConfiguration config, Label requiredLabel) {
+    private synchronized ComputeEngineInstance getAnAgent(InstanceConfiguration config, String labels) {
```
Comment (craigatgoogle) [inline src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineCloud.java]: Is this method dead code? It doesn't seem to be called from anywhere.

File: src/main/java/com/google/jenkins/plugins/computeengine/InstanceConfiguration.java [hunk 2]
```diff
@@ -225,13 +225,13 @@ public void appendLabel(String key, String value) {
         googleLabels.put(key, value);
     }
 
-    public ComputeEngineInstance provision(TaskListener listener, Label requiredLabel) throws IOException {
+    public ComputeEngineInstance provision(TaskListener listener, String labels) throws IOException {
```
Comment (craigatgoogle) [inline src/main/java/com/google/jenkins/plugins/computeengine/InstanceConfiguration.java]: I'm nervous about having a single string parameter which represents multiple values. This has format assumptions baked into it, which need to be extended to any callers. Consider instead using an ImmutableList<String> param and encapsulate the string formatting assumptions as an implementation detail.

Comment (craigatgoogle) [reply src/main/java/com/google/jenkins/plugins/computeengine/InstanceConfiguration.java]: Also would like to see a unit test which ensures no regressions.

**Comment by ingwarsw** [comment]
@rachely3n @craigatgoogle Actually Im changing template pr #12 to add template labels on top of internal needed labels.. 
After that it will cover all functionality of that PR..

Do you think it will be still needed?

File: src/main/java/com/google/jenkins/plugins/computeengine/InstanceConfiguration.java (hunk 2 above)
Comment (ingwarsw) [reply src/main/java/com/google/jenkins/plugins/computeengine/InstanceConfiguration.java]: I used that PR a bit before it was needed for my setup, and it worked with plain labels..

Under the hood they are also setting `string` labels so its rather ok..
And we would just take string split it by ` ` and then join it by ` `..
We cannot du much more I think...

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineCloud.java (hunk 1 above)
Comment (ingwarsw) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineCloud.java]: Seems that way..

But it was there previously so it should not be removed in this PR but rather extra PR for cleaning..

File: src/main/java/com/google/jenkins/plugins/computeengine/InstanceConfiguration.java (hunk 2 above)
Comment (craigatgoogle) [reply src/main/java/com/google/jenkins/plugins/computeengine/InstanceConfiguration.java]: I would rather decrease any such tight-coupling. String splitting/joining is still preferable than having to maintain assumptions about the format of data in multiple places. Making the provision method take a list makes it a much simpler method.

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineCloud.java (hunk 1 above)
Comment (craigatgoogle) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineCloud.java]: In that case, an issue for the re-factor should be created and linked from here in a comment.

**Comment by craigatgoogle** [comment]
This would be a good candidate for an integration test which uses multiple labels.

**Comment by rachely3n** [comment]
@Crazybus I will be closing this PR since I merged it here https://github.com/jenkinsci/google-compute-engine-plugin/pull/15. Had to add some integration tests/fix merge conflicts.

Thanks for contributing! This was a good catch.

**Pull request closed without merging**